### ISSUE TITLE
POST is writing Content-Lentgth header twice

### ISSUE THREAD

**Issue by lardcanoe** [socketry/http-protocol#2]
First off, all the `async` gems look really promising. Great work so far.

I'm doing a simple POST using a regular string as the body and the `Content-Type` header is getting written twice.

I hacked up `write_headers` to skip it in lib/http/protocol/http1/connection.rb and then have `write_fixed_length_body` be responsible for writing it out. This got me moving.

My usage is with the aws-sdk gem. I make 15 concurrent calls to all the AWS regions, so i figured using the `async` suite of libraries would ultimately be faster than spawning a bunch of threads to do the requests, but the threads were 3x faster after dozens of tests. Wondering if you have a good example of how to write clients to prove out a usage example like this.

**Comment by ioquatix** [comment]
Basically, you shouldn't explicitly specify the `content-length` header. Can you show me a simple example of what you are trying to do?

If you want to make 15 concurrent calls, you should spawn 15 tasks. Did you do that?

**Comment by lardcanoe** [comment]
To confirm I am using it right, without any http:

```
::Async::Reactor.run do |task|
  15.times do |x|
    task.async do |_nested_task|
      puts "#{x}: Start..."
      _nested_task.sleep 5 * rand
      puts "#{x}: End."
    end
  end
  puts 'Waiting on tasks to finish...'
  task.children.each(&:wait)
end.wait
```

Results in:

```
➜  async-post 9.2.0.0 k8s:(minikube)  jruby -G post.rb
OpenSSL implementation doesn't support ALPN.
0: Start...
1: Start...
2: Start...
3: Start...
4: Start...
5: Start...
6: Start...
7: Start...
8: Start...
9: Start...
10: Start...
11: Start...
12: Start...
13: Start...
14: Start...
Waiting on tasks to finish...
10: End.
14: End.
11: End.
12: End.
1: End.
0: End.
8: End.
3: End.
13: End.
6: End.
5: End.
4: End.
7: End.
9: End.
2: End.
D, [2018-10-15T13:25:33.077483 #15305] DEBUG -- <Async::Reactor:0x7d0 stopped=false>: Exiting run-loop because finished.
```

Based on that, I think I am constructing the code properly. You mentioed to spawn 15 tasks, but I have 1 task with 15 children. Is that the wrong approach?

**Comment by lardcanoe** [comment]
```
::Async::Reactor.run do |task|
  tasks = 15.times.collect do |x|
    task.async do
      puts "#{x}: Start..."
      headers = {}
      endpoint = ::Async::HTTP::URLEndpoint.parse("http://localhost:8880", reuse_port: true)
      client = ::Async::HTTP::Client.new(endpoint, Async::HTTP::Protocol::HTTP11)
      async_resp = client.post "/hello", headers, [x.to_s]
      puts "#{x}: End."
    end
  end
  puts 'Waiting on tasks to finish...'
  tasks.each(&:wait)
end.wait
```

Results in:

```
0: Start...
D, [2018-10-15T13:31:45.757699 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.763610 #19199] DEBUG -- #<Async::HTTP::Client:0x487ba080>: Making connection to #<Async::HTTP::URLEndpoint:0x167a6ea5 @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.802198 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
1: Start...
D, [2018-10-15T13:31:45.825834 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.826691 #19199] DEBUG -- #<Async::HTTP::Client:0x61cf6b8c>: Making connection to #<Async::HTTP::URLEndpoint:0x52cda30e @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.827961 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
2: Start...
D, [2018-10-15T13:31:45.835008 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.836601 #19199] DEBUG -- #<Async::HTTP::Client:0x2c41ebac>: Making connection to #<Async::HTTP::URLEndpoint:0x50840b3f @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.837832 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
3: Start...
D, [2018-10-15T13:31:45.842353 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.843270 #19199] DEBUG -- #<Async::HTTP::Client:0x318f3283>: Making connection to #<Async::HTTP::URLEndpoint:0x7937b46d @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.844281 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
4: Start...
D, [2018-10-15T13:31:45.848939 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.849789 #19199] DEBUG -- #<Async::HTTP::Client:0x40cf370b>: Making connection to #<Async::HTTP::URLEndpoint:0x4b09841a @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.850992 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
5: Start...
D, [2018-10-15T13:31:45.853739 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.854629 #19199] DEBUG -- #<Async::HTTP::Client:0x2bf20edd>: Making connection to #<Async::HTTP::URLEndpoint:0x29b4e28c @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.856725 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
6: Start...
D, [2018-10-15T13:31:45.860421 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.861450 #19199] DEBUG -- #<Async::HTTP::Client:0x45a3a63f>: Making connection to #<Async::HTTP::URLEndpoint:0x3292028 @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.863395 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
7: Start...
D, [2018-10-15T13:31:45.870807 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.871671 #19199] DEBUG -- #<Async::HTTP::Client:0x341524e3>: Making connection to #<Async::HTTP::URLEndpoint:0x5da0434 @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.873162 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
8: Start...
D, [2018-10-15T13:31:45.877557 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.878435 #19199] DEBUG -- #<Async::HTTP::Client:0x7c1e3004>: Making connection to #<Async::HTTP::URLEndpoint:0x1ebe030b @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.880676 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
9: Start...
D, [2018-10-15T13:31:45.887468 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.888117 #19199] DEBUG -- #<Async::HTTP::Client:0x1b861196>: Making connection to #<Async::HTTP::URLEndpoint:0x3422f9a6 @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.889021 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
10: Start...
D, [2018-10-15T13:31:45.892259 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.893158 #19199] DEBUG -- #<Async::HTTP::Client:0x58c8301>: Making connection to #<Async::HTTP::URLEndpoint:0x5991368d @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.894289 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
11: Start...
D, [2018-10-15T13:31:45.897623 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.898326 #19199] DEBUG -- #<Async::HTTP::Client:0x4deb5517>: Making connection to #<Async::HTTP::URLEndpoint:0x128386e @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.899374 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
12: Start...
D, [2018-10-15T13:31:45.902403 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.903219 #19199] DEBUG -- #<Async::HTTP::Client:0x4249f702>: Making connection to #<Async::HTTP::URLEndpoint:0x4719d511 @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.904376 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
13: Start...
D, [2018-10-15T13:31:45.908457 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.909159 #19199] DEBUG -- #<Async::HTTP::Client:0x38c3ce38>: Making connection to #<Async::HTTP::URLEndpoint:0x61069fdf @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.910760 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
14: Start...
D, [2018-10-15T13:31:45.914816 #19199] DEBUG -- #<Async::HTTP::Pool resources= limit=>: No resources resources, allocating new one...
D, [2018-10-15T13:31:45.915422 #19199] DEBUG -- #<Async::HTTP::Client:0x58e8ac81>: Making connection to #<Async::HTTP::URLEndpoint:0x649a9609 @url=#<URI::HTTP http://localhost:8880/>, @options={:reuse_port=>true}, @endpoint=nil>
D, [2018-10-15T13:31:45.916164 #19199] DEBUG -- Async::IO::Socket: Connecting to #<Addrinfo: 127.0.0.1:8880>
Waiting on tasks to finish...
D, [2018-10-15T13:31:45.921408 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x3d7e13e6>
D, [2018-10-15T13:31:45.922163 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x3d7e13e6>: POST /hello {}
D, [2018-10-15T13:31:45.925094 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x1ad4d6da>
D, [2018-10-15T13:31:45.925782 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x1ad4d6da>: POST /hello {}
D, [2018-10-15T13:31:45.927951 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x3447b69>
D, [2018-10-15T13:31:45.928715 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x3447b69>: POST /hello {}
D, [2018-10-15T13:31:45.947528 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x52c81af3>
D, [2018-10-15T13:31:45.948488 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x52c81af3>: POST /hello {}
D, [2018-10-15T13:31:45.950565 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x34461077>
D, [2018-10-15T13:31:45.952843 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x34461077>: POST /hello {}
D, [2018-10-15T13:31:45.956351 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x679cdb75>
D, [2018-10-15T13:31:45.957140 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x679cdb75>: POST /hello {}
D, [2018-10-15T13:31:45.965538 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x5f4f476a>
D, [2018-10-15T13:31:45.972171 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x5f4f476a>: POST /hello {}
D, [2018-10-15T13:31:45.975410 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x2b5fea0c>
D, [2018-10-15T13:31:45.976701 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x2b5fea0c>: POST /hello {}
D, [2018-10-15T13:31:45.978895 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x349b311f>
D, [2018-10-15T13:31:45.979670 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x349b311f>: POST /hello {}
D, [2018-10-15T13:31:45.982136 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x24930455>
D, [2018-10-15T13:31:45.984990 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x24930455>: POST /hello {}
D, [2018-10-15T13:31:45.989460 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x411bdbc9>
D, [2018-10-15T13:31:45.990176 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x411bdbc9>: POST /hello {}
D, [2018-10-15T13:31:45.992128 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x21420645>
D, [2018-10-15T13:31:45.992777 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x21420645>: POST /hello {}
D, [2018-10-15T13:31:45.994661 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x13a97593>
D, [2018-10-15T13:31:45.995333 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x13a97593>: POST /hello {}
D, [2018-10-15T13:31:45.997263 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x37c63f5>
D, [2018-10-15T13:31:45.997953 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x37c63f5>: POST /hello {}
D, [2018-10-15T13:31:46.004513 #19199] DEBUG -- #<Async::HTTP::Pool resources=1/1 limit=>: Wait for resource #<Async::HTTP::Protocol::HTTP1::Client:0x6849ef7c>
D, [2018-10-15T13:31:46.008871 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x6849ef7c>: POST /hello {}
D, [2018-10-15T13:31:46.225636 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x3447b69>: HTTP/1.1 200 OK
3: End.
D, [2018-10-15T13:31:46.239709 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x3d7e13e6>: HTTP/1.1 200 OK
2: End.
D, [2018-10-15T13:31:46.257338 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x52c81af3>: HTTP/1.1 200 OK
14: End.
D, [2018-10-15T13:31:46.259506 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x679cdb75>: HTTP/1.1 200 OK
11: End.
D, [2018-10-15T13:31:46.274982 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x1ad4d6da>: HTTP/1.1 200 OK
10: End.
D, [2018-10-15T13:31:46.299872 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x34461077>: HTTP/1.1 200 OK
6: End.
D, [2018-10-15T13:31:46.341451 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x2b5fea0c>: HTTP/1.1 200 OK
7: End.
D, [2018-10-15T13:31:46.352409 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x5f4f476a>: HTTP/1.1 200 OK
1: End.
D, [2018-10-15T13:31:46.358608 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x13a97593>: HTTP/1.1 200 OK
4: End.
D, [2018-10-15T13:31:46.390547 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x411bdbc9>: HTTP/1.1 200 OK
12: End.
D, [2018-10-15T13:31:46.394509 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x37c63f5>: HTTP/1.1 200 OK
0: End.
D, [2018-10-15T13:31:46.399853 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x24930455>: HTTP/1.1 200 OK
9: End.
D, [2018-10-15T13:31:46.404437 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x21420645>: HTTP/1.1 200 OK
13: End.
D, [2018-10-15T13:31:46.417133 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x349b311f>: HTTP/1.1 200 OK
8: End.
D, [2018-10-15T13:31:46.420536 #19199] DEBUG -- #<Async::HTTP::Protocol::HTTP1::Client:0x6849ef7c>: HTTP/1.1 200 OK
5: End.
D, [2018-10-15T13:31:46.422412 #19199] DEBUG -- <Async::Reactor:0x7d0 stopped=false>: Exiting run-loop because finished.
```

I think the double content-length was my fault. I was passing it in from the aws-sdk gem without realizing it. Will validate that.

**Comment by ioquatix** [comment]
So:

- Debug logging adds overhead, so make sure that's off if you are testing performance.
- Reuse clients if possible.
- Read/finish/close the response body e.g. `async_resp.read`.
- Close the client when done `client.close`
- Cache the client where possible unless you actually want multiple connections. The client has a connection pool which will limit and reuse connections to the remote endpoint.

```ruby
clients = {}

endpoint = Async::HTTP::URLEndpoint.parse("...")

clients[endpoint] ||= Async::HTTP::Client.new(endpoint, ...)

clients.each_value(&:close)
clients.clear
```

Here is a `Proxy` middleware which does roughly the above: https://github.com/socketry/falcon/blob/master/lib/falcon/proxy.rb

**Comment by ioquatix** [comment]
Also, I'd be interested to see any comparisons you do and understand how you are using it with `aws`.

**Comment by ioquatix** [comment]
In order to simplify this, I added `Async::HTTP::Internet` which lets you make requests with full URLs, and it handles the multiplexing of connections.

Feel free to open a new issue if your original problem is not resolved.
